python/ray/serve/async_goal_manager.py

```diff
@@ -8,34 +8,92 @@
 
 
 class AsyncGoalManager:
+    """
+    Helper class to facilitate ServeController's async goal creation, tracking
+    and termination by maintaining a dictionary of <goal_id, asyncio.Future>.
+
+    Ray serve external api calls can be assigned with a goal_id for tracking,
+    and can be either awaited as blocking call or delegate to run
+    asynchronously.
+
+    Ex:
+        deploy() -> goal_id = UUID('1b985345-f0ae-4cb7-8349-aa2ea881bf89')
+            if blocking:
+                # Block until deploy() action is done
+                wait_for_goal(goal_id)
+            else:
+                # Immediately return and let controller to deploy async
+                return goal_id
+
+    As of now, each external api call as well as goal states are also
+    checkpointed in case of serve controller failure.
+    """
+
     def __init__(self):
-        self._pending_goals: Dict[GoalId, asyncio.Event] = dict()
+        self._pending_goals: Dict[GoalId, asyncio.Future] = dict()
 
     def num_pending_goals(self) -> int:
         return len(self._pending_goals)
 
     def create_goal(self, goal_id: Optional[GoalId] = None) -> GoalId:
+        """
+        Create a new asyncio.Future as goal and return its id. If an id
+        is already given, assign a future to it and return same id.
+        """
         if goal_id is None:
             goal_id = uuid4()
-        self._pending_goals[goal_id] = asyncio.Event()
+
+        self._pending_goals[
+            goal_id] = asyncio.get_running_loop().create_future()
+
         return goal_id
 
-    def complete_goal(self, goal_id: GoalId) -> None:
+    def complete_goal(self,
+                      goal_id: GoalId,
+                      exception: Optional[Exception] = None) -> None:
+        """
+        Mark given goal as completed, if an exception object is provided,
+        set exception instead.
+        """
         logger.debug(f"Completing goal {goal_id}")
-        event = self._pending_goals.pop(goal_id, None)
-        if event:
-            event.set()
+        future = self._pending_goals.pop(goal_id, None)
+        if future:
+            if exception:
+                future.set_exception(exception)
+            else:
+                future.set_result(goal_id)
 
     def check_complete(self, goal_id: GoalId) -> bool:
-        return goal_id not in self._pending_goals
+        """
+        Check if given goal is completed.
+        """
+        if goal_id not in self._pending_goals:
+            return True
```

Comment:
maybe add a comment here that if the goal previously errored but was removed, the exception won't be re-raised (minor edge case)